valhalla/skadi/sample.h

```diff
@@ -56,30 +63,81 @@ class sample {
    */
   static std::string get_hgt_file_name(uint16_t index);
 
+  enum class format_t { UNKNOWN = 0, RAW = 1, GZIP = 2 };
+
+  class tile_data {
+  private:
+    sample* s;
+    const int16_t* data;
+    uint16_t index;
+    bool reusable;
+
+  public:
+    tile_data(sample* s, uint16_t index, bool reusable, const int16_t* data);
+    tile_data(const tile_data& other);
+    ~tile_data();
+    //      tile_data& operator= (tile_data &other);
+    tile_data& operator=(tile_data&& other);
+
+    inline operator bool() const {
+      return data != nullptr;
+    }
+    double get(double u, double v);
+  };
+
   /**
    * @param  index  the index of the data tile being requested
    * @return the array of data or nullptr if there was none
    */
-  const int16_t* source(uint16_t index) const;
-
-  enum class format_t { UNKNOWN = 0, GZIP = 1, RAW = 3 };
-  /**
-   * maps a new source, used at start up and called periodically
-   * for lazily loaded sources
-   *
-   * @param  index  the index of the data tile being mapped
-   * @param  format the format of the data tile being mapped
-   * @param  file   the file name of the data tile being mapped
-   */
-  void map(uint16_t index, format_t format, const std::string& file);
+  sample::tile_data source(uint16_t index);
+
+  void increment_usages(uint16_t index);
+  void decrement_usages(uint16_t index);
+
+  class cache_item_t {
+  private:
+    format_t format;
+    midgard::mem_map<char> data;
+    int usages;
+    const char* unpacked;
+
+  public:
+    cache_item_t() : format(format_t::UNKNOWN), usages(0), unpacked(nullptr) {
+    }
+    cache_item_t(cache_item_t&&) = default;
+    ~cache_item_t() {
+      free((void*)unpacked);
+    }
+    bool init(const std::string& file, format_t format);
+
+    inline const char* get_data() const {
+      return data.get();
+    }
+    inline format_t get_format() const {
+      return format;
+    }
+    inline int& get_usages() {
+      return usages;
+    }
+    inline const char* get_unpacked() {
+      return unpacked;
+    }
+    inline const char* detach_unpacked() {
+      auto rv = unpacked;
+      unpacked = nullptr;
+      return rv;
+    }
+
+    bool unpack(const char* unpacked);
+    static boost::optional<std::pair<uint16_t, sample::format_t>>
+    parse_hgt_name(const std::string& name);
+  };
 
   // using memory maps
-  mutable std::vector<std::pair<format_t, midgard::mem_map<char>>> mapped_cache;
```

Comment:
so this is the main question to me. completely removing memory mapping will have some performance impact. i dont care as much about peformance impact for offline processing like that we have in `valhalla_build_tiles` but when we run the service and somehow does a `/height` request it needs to be super fast (as it was before). with a naive caching scheme we will not be able to beat what the OS is doing with memory mapping but we'd have to test it to find out. indeed i know that memory mapping many small files can be slower than just reading them from disk and in that scenario its faster to have an in memory cache like we do for graph tiles but there we are talking about 10s of kilobytes and here we are talking about tiles that are 25 megabytes in size. anyway we need to consider how this change affects the service. it could be for the better (ie we load up the 300GB of compressed elevation data and now its fast enough to be usable but still not need a 1.6TB disk). i just know from 3 years ago when i implemented lz4 it ended up not being fast enough performance wise though i will say i dont remember what my caching looked like (though i guess git history would tell us).